### PR TITLE
BM 09/13/20 - Fetch Recipes And Ingredients

### LINKED ISSUE #9: User should instantiate Cookbook, cookbook should instantiate a Recipe

### LINKED ISSUE #24: Fetch - Retrieve all data from an endpoint (GET)
* wcUsersData: https://fe-apps.herokuapp.com/api/v1/whats-cookin/1911/users/wcUsersData

* ingredientsData: https://fe-apps.herokuapp.com/api/v1/whats-cookin/1911/ingredients/ingredientsData

* recipesData: https://fe-apps.herokuapp.com/api/v1/whats-cookin/1911/recipes/recipeData

### LINKED ISSUE #26: Fetch - Error Handling (catch)
handle the case where they submit their data and an error message returns from the request, and also validate the input fields on the client-side

### LINKED ISSUE #45: Move all querySelectors out of functions into global of scripts.js

### PR DESCRIPTION
**Pull request by bretm9**
## Description
* The application now fetches data for recipeData and IngredientData onload, and renders cards and recipe windows using fetched data. 
* Refactored populateCards to incorporate dipslayFavorites functionality
* removed displayFavorites

## Context
Added fetch functionality in order to pull down recipe and ingredient data from the server rather than live locally.

## Where To Start
Most changes were made in scripts.js, with a few tweaks to dom-updates.js

## Affected areas of application
List all parts of the website that will be altered from this PR.

## How Has This Been Tested?

- [x] Chai / Mocha
- [x] Chrome Browser / inspector
- [x] .catch error-handlers

## Relevant Tickets
Closes #26, 
closes #24, 
closes #45,
closes #9

## Questions or Notes (if applicable):
We might want to refactor the fetch functionality in the future because there are two places where we're fetching the same data (onload in scripts.js and in dom-updates.js).

### CODE REVIEW

File: src/scripts.js [hunk 1]
```diff
@@ -67,69 +67,42 @@ function cardButtonConditionals(event) {
...
+    domUpdates.removeAll(cardArea);
   }
   if (!user.favoriteRecipes.length) {
-  domUpdates.showNoFavorites();
-    domUpdates.populateCards(cookbook.recipes);
+  domUpdates.showNoFavorites(favButton);
+    domUpdates.populateCards(recipeData, cardArea);
     return
     // we can use break if we are not trying to return anything
   } else {
```
Comment (Chriscastanuela) [inline src/scripts.js:83]: I'm putting a pin in this: "refactor to shorten (if we have the time) if statements with ternary operator"

File: test/dom-updates-test.js [hunk 2]
```diff
@@ -1,16 +1,14 @@
 const chai = require('chai');
 const expect = chai.expect;
 const spies = require('chai-spies');
-import DomUpdates from '../src/dom-updates';
+import domUpdates from '../src/dom-updates';
```
Comment (Chriscastanuela) [inline test/dom-updates-test.js:4]: Great catch since we changed domUpdates from a class to an object literal

**Review by Chriscastanuela — APPROVED** [review]

File: src/scripts.js [hunk 3]
```diff
@@ -86,30 +86,30 @@ function viewFavorites() {
 }
 
 function favoriteCard(event) {
-  let targetedID;
-  if (event.target.id.includes('-')) {
-    targetedID = event.target.id.slice(0, event.target.id.indexOf('-'))
+  // let targetedID;
+  // if (event.target.id.includes('-')) {
+    let targetedID = event.target.id.slice(0, event.target.id.indexOf('-'));
```
Comment (holladayian) [inline src/scripts.js]: I'm glad we got this cleaned up. It makes the function nicer to look at. I'm still not sure that every event.target.id.includes('-'), but I suppose we'll see when we test with the fetched data.

File: src/scripts.js [hunk 4]
```diff
@@ -86,30 +86,30 @@ function viewFavorites() {
...
+
+  if (!target.classList.contains('favorite-active')) {
     user.addToCategory(specificRecipe, "favoriteRecipes");
-  } else if (event.target.classList.contains('favorite-active')) {
-    event.target.classList.remove('favorite-active');
-    user.removeFromCategory(specificRecipe, "favoriteRecipes")
+    domUpdates.favoritesToggle();
+  } else if (target.classList.contains('favorite-active')) {
+    user.removeFromCategory(specificRecipe, "favoriteRecipes");
+    domUpdates.favoritesToggle();
```
Comment (holladayian) [inline src/scripts.js]: Sweet job splitting this functionality into DOM updates. I know it's technically longer, but I think it does a good job of separating out different functionalities of code.

**Review by holladayian — APPROVED** [review]
Good job! Just some minor things to note, but nothing to fix.

File: src/scripts.js [hunk 5]
```diff
@@ -1,24 +1,27 @@
...
 let favButton = document.querySelector('.view-favorites');
 let cardArea = document.querySelector('.all-cards');
 
+let user, pantry, newUser, recipeData, ingredientsData;
 
-let cookbook = new Cookbook(recipeData);
-let user, pantry, newUser;
-
-window.onload = onStartup();
+window.onload =
```
Comment (holladayian) [inline src/scripts.js:15]: Nice thinking, slapping this into an onload.

File: src/scripts.js [hunk 6]
```diff
@@ -1,24 +1,27 @@
...
 import './css/styles.scss';
 
-import recipeData from './data/recipes';
-import ingredientsData from './data/ingredients';
-
 import Pantry from './pantry';
 import Recipe from './recipe';
 import User from './user';
-import Cookbook from './cookbook';
 import domUpdates from './dom-updates';
```
Comment (holladayian) [inline src/scripts.js]: Do we need to import at all if the what we're importing is already hooked up through webpack?

File: src/scripts.js [hunk 7]
```diff
@@ -39,7 +42,7 @@ function fetchUserData() {
   });
   user = new User(userId, newUser.name, newUser.pantry);
   pantry = new Pantry(newUser.pantry);
-  domUpdates.populateCards(cookbook.recipes, cardArea, user.favoriteRecipes);
+  domUpdates.populateCards(recipeData, cardArea, user.favoriteRecipes);
```
Comment (holladayian) [inline src/scripts.js:46]: We still have access to cookbook.recipes, correct? I'm guessing you tested this and recipeData works just as well, so just a question, no action needed.

**Review by holladayian — APPROVED** [review]
Just a few questions for clarification. It seems that everything works properly, so questions are purely scholarly.

File: src/dom-updates.js [hunk 8]
```diff
@@ -15,33 +15,46 @@ let domUpdates = {
...
     if (cardArea.classList.contains('all')) {
       cardArea.classList.remove('all')
     }
-    recipes.forEach(recipe => {
-      let isFavorite;
-      if (userFavorites && userFavorites.find(favoriteRecipe => favoriteRecipe.id === recipe.id)) {
-        isFavorite = 'favorite-active';
-      } else {
-        isFavorite = '';
-      }
```
Comment (holladayian) [inline src/dom-updates.js]: I think we'll be able to keep this piece for adding to recipe, however the filtering with userFavorites will have to be a desired list -i.e. userFavorites or recipesToCook, I think

**Review by holladayian — APPROVED** [review]
Something to note when we add functionality to filter through recipesToCook.

**Comment by holladayian** [comment]
Good work! Mainly just some minor questions and refactoring. Only "larger" thing to note is how to tackle functionality of showing recipesToCook.

**Pull request merged**